### PR TITLE
WIP: Add support for the "interface info" mode of the /v2/interfaces API

### PR DESCRIPTION
**Pull request by jhenstridge**
This is still in progress, but I thought I'd file the pull request early.

There are two modes of output for the `GET /v2/interfaces` API: the current one returning the list of all connections, and a second mode when the `select` query argument is passed, which shows information about interface types themselves.

This isn't a case of one output format replacing the other, so I haven't gone with a `foo2` style name for the new APIs (in the Go client API, these two modes are exposed as `Connections()` and `Interfaces()`).  I've reused the `SnapdPlug` and `SnapdSlot` types in the new APIs, since the Go client uses the same types too.

If you've got preferences for different naming for anything I've added, I can easily change things.

The branch is not quite finished though:

- [x] Implement C API
- [x] Add support to mock-snapd, and test new API
- [ ] Implement Qt API, and tests

### CODE REVIEW

**Comment by jhenstridge** [comment]
So, I've got this mostly working for the glib side, including GIR bindings:

    >>> from gi.repository import Snapd
    >>> c = Snapd.Client()
    >>> ifaces = c.get_interface_info_sync(['content'], Snapd.InterfaceFlags.INCLUDE_PLUGS | Snapd.InterfaceFlags.INCLUDE_SLOTS)
    >>> content = ifaces[0]
    >>> content.get_name()
    'content'
    >>> content.get_summary()
    'allows sharing code and data with other snaps'
    >>> plug = content.get_plugs()[0]
    >>> plug.get_snap()
    'file-roller'
    >>> plug.get_name()
    'mir-libs'
    >>> plug.get_attribute('default-provider')
    GLib.Variant('s', 'mir-libs')

I'd like some feedback on whether the API looks okay before trying to expand this to the Qt side.

File: snapd-glib/requests/snapd-get-interfaces.c [hunk 1]
```diff
@@ -263,77 +68,13 @@ parse_get_interfaces_response (SnapdRequest *request, SoupMessage *message, GErr
...
-                             "interface", _snapd_json_get_string (object, "interface", NULL),
-                             "label", _snapd_json_get_string (object, "label", NULL),
-                             "connections", connections,
-                             "attributes", attributes,
-                             // FIXME: apps
-                             NULL);
-        g_ptr_array_add (plug_array, g_steal_pointer (&plug));
-    }
+    plug_array = _snapd_json_parse_plug_array (plugs, error);
+    if (!plug_array)
```
Comment (robert-ancell) [inline snapd-glib/requests/snapd-get-interfaces.c]: I've used the style plug_array == NULL for pointers in snapd-glib.

File: snapd-glib/requests/snapd-get-interfaces.c [hunk 2]
```diff
@@ -263,77 +68,13 @@ parse_get_interfaces_response (SnapdRequest *request, SoupMessage *message, GErr
...
-                             "interface", _snapd_json_get_string (object, "interface", NULL),
-                             "label", _snapd_json_get_string (object, "label", NULL),
-                             "connections", connections,
-                             "attributes", attributes,
-                             // FIXME: apps
-                             NULL);
-        g_ptr_array_add (slot_array, g_steal_pointer (&slot));
-    }
+    slot_array = _snapd_json_parse_slot_array (slots, error);
+    if (!slot_array)
```
Comment (robert-ancell) [inline snapd-glib/requests/snapd-get-interfaces.c]: I've used the style slot_array == NULL for pointers in snapd-glib.

File: snapd-glib/snapd-client-sync.c [hunk 3]
```diff
@@ -519,6 +519,36 @@ snapd_client_get_interfaces_sync (SnapdClient *client,
     return snapd_client_get_interfaces_finish (client, data.result, plugs, slots, error);
 }
 
+/**
+ * snapd_client_get_interface_info_sync:
+ * @client: a #SnapdClient.
+ * @names: (array zero-terminated=1): a null-terminated array of interface names or %NULL.
```
Comment (robert-ancell) [inline snapd-glib/snapd-client-sync.c]: Missing (allow-none). Also if NULL is provided what happens? Should it say ".. or %NULL for all interfaces"?

File: snapd-glib/snapd-client-sync.c [hunk 4]
```diff
@@ -519,6 +519,36 @@ snapd_client_get_interfaces_sync (SnapdClient *client,
...
+ * @flags: control what information is returned about the interfaces.
+ * @cancellable: (allow-none): a #GCancellable or %NULL.
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL to ignore.
+ *
+ * Get the installed snap interfaces.
```
Comment (robert-ancell) [inline snapd-glib/snapd-client-sync.c]: Could you add a comment comparing this to get_interfaces () and one in get_interfaces () that does the same? In particular I'd like to steer users to choosing the correct API given they have such similar names.

File: snapd-glib/snapd-client-sync.c [hunk 5]
```diff
@@ -519,6 +519,36 @@ snapd_client_get_interfaces_sync (SnapdClient *client,
...
+ */
+GPtrArray *
+snapd_client_get_interface_info_sync (SnapdClient *client,
+                                      gchar **names,
+                                      SnapdInterfaceFlags flags,
+                                      GCancellable *cancellable, GError **error)
+{
+    g_auto(SyncData) data = { 0 };
+
+    g_return_val_if_fail (SNAPD_IS_CLIENT (client), FALSE);
```
Comment (robert-ancell) [inline snapd-glib/snapd-client-sync.c]: FALSE -> NULL

File: snapd-glib/snapd-client.c [hunk 6]
```diff
@@ -1870,6 +1871,75 @@ snapd_client_get_interfaces_finish (SnapdClient *client, GAsyncResult *result,
     return TRUE;
 }
 
+/**
+ * snapd_client_get_interface_info_async:
+ * @client: a #SnapdClient.
+ * @names: (array zero-terminated=1): a null-terminated array of interface names or %NULL.
```
Comment (robert-ancell) [inline snapd-glib/snapd-client.c]: Same changes as for the _sync call.

File: snapd-glib/snapd-client-sync.c [hunk 7]
```diff
@@ -519,6 +519,36 @@ snapd_client_get_interfaces_sync (SnapdClient *client,
     return snapd_client_get_interfaces_finish (client, data.result, plugs, slots, error);
 }
 
+/**
+ * snapd_client_get_interface_info_sync:
+ * @client: a #SnapdClient.
+ * @names: (array zero-terminated=1): a null-terminated array of interface names or %NULL.
+ * @flags: control what information is returned about the interfaces.
```
Comment (robert-ancell) [inline snapd-glib/snapd-client-sync.c]: Use the same form as other flags:

`@flags: a set of #SnapdInterfaceFlags to control what information is returned about the interfaces.`

That way there is an easy link to the flags in the documentation.

File: snapd-glib/snapd-client.h [hunk 8]
```diff
@@ -150,6 +150,27 @@ typedef enum
...
+ *
+ * Since: 1.42
+ */
+typedef enum
+{
+    SNAPD_INTERFACE_FLAGS_NONE = 0,
+    SNAPD_INTERFACE_FLAGS_INCLUDE_DOCS = 1 << 0,
+    SNAPD_INTERFACE_FLAGS_INCLUDE_PLUGS = 1 << 1,
+    SNAPD_INTERFACE_FLAGS_INCLUDE_SLOTS = 1 << 2,
+    SNAPD_INTERFACE_FLAGS_ONLY_CONNECTED = 1 << 3,
```
Comment (robert-ancell) [inline snapd-glib/snapd-client.h]: Please align values like the other flags.

File: snapd-glib/snapd-client.c [hunk 9]
```diff
@@ -1870,6 +1871,75 @@ snapd_client_get_interfaces_finish (SnapdClient *client, GAsyncResult *result,
...
+ * @flags: control what information is returned about the interfaces.
```
Comment (robert-ancell) [inline snapd-glib/snapd-client.c]: Same changes as for the _sync call.

File: tests/test-glib.c [hunk 10]
```diff
@@ -2576,6 +2576,180 @@ test_get_interfaces_no_snaps (void)
...
+    g_assert_cmpint (slots->len, ==, 1);
+
+    slot = slots->pdata[0];
+    g_assert_cmpstr (snapd_slot_get_name (slot), ==, "slot1");
+    g_assert_cmpstr (snapd_slot_get_snap (slot), ==, "snap1");
+
+    /* If we select connected interfaces, then we don't see the
+     * unconnected plug */
+    g_clear_pointer (&ifaces, g_ptr_array_unref);
+    ifaces = snapd_client_get_interface_info_sync (client, NULL, SNAPD_INTERFACE_FLAGS_INCLUDE_DOCS | SNAPD_INTERFACE_FLAGS_INCLUDE_PLUGS | SNAPD_INTERFACE_FLAGS_INCLUDE_SLOTS | SNAPD_INTERFACE_FLAGS_ONLY_CONNECTED, NULL, &error);
```
Comment (robert-ancell) [inline tests/test-glib.c]: For the tests I've done one snapd-glib call per test - so ideally this should be split into multiple tests. I'm not going to block on that if you don't want to make the change.

**Comment by robert-ancell** [comment]
Thanks for this! I noticed this before I went away, so glad someone else is doing the work :)

I added comments on the changes, just small details to fix. The names are very close together, but I don't think there's an easy way to differentiate them. In particular, I think we just need the docs to be clear in steering the developer to understand which API call to use.

**Comment by robert-ancell** [comment]
We should also update https://github.com/snapcore/snapd/wiki/REST-API with this API

File: snapd-glib/snapd-client-sync.c [hunk 11]
```diff
@@ -519,6 +519,36 @@ snapd_client_get_interfaces_sync (SnapdClient *client,
...
+ * @cancellable: (allow-none): a #GCancellable or %NULL.
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL to ignore.
+ *
+ * Get the installed snap interfaces.
+ *
+ * Returns: (transfer container) (element-type SnapdInterface): the interfaces.
+ *
+ * Since: 1.42
```
Comment (robert-ancell) [inline snapd-glib/snapd-client-sync.c]: Needs to be since 1.43 now (and in other places).

**Comment by robert-ancell** [comment]
I'm working on updating this to apply in https://github.com/robert-ancell/snapd-glib/tree/interface-info-update

**Comment by robert-ancell** [comment]
See https://github.com/snapcore/snapd-glib/pull/64

**Pull request closed without merging**